### PR TITLE
Wrap compilation in a cross-os-process lock

### PR DESCRIPTION
**Pull request by jonatanklosko**
This wraps mix compile in a lock, such that only one compilation is running at a time. This helps to ensure that no duplicated compilation work is being done for the given target.

More specifically, we wrap tasks such as `compile.all`, `compile.elixir`, `compile.erlang`, `compile.deps` in the lock. Technically wrapping `compile.all` should be enough, but child tasks (such as `compile.elixir`) can still be invoked directly, hence we wrap each of them in the lock (lock calls can be nested safely).

As for the lock implementation, initially we planned to use Unix domain sockets, but it turned out to be non-ideal. Currently the `:socket` implementation on Windows has an important bug (but it's likely to be patched soon). The main reason we turned down this approach, though, is that the socket path is limited to 104 characters. This makes it too brittle, because even the temporary directory can be long (on macOS or via custom TMPDIR), and combined with additional naming we would need, we would be close to the limit.

Instead, the current implementation relies on TCP sockets and hard links. The module has a detailed comment outlining the algorithm. For the record, both implementations can be found in [jonatanklosko/elixir_lock](https://github.com/jonatanklosko/elixir_lock). The implementation in this PR has a different public API, tailored for the Mix use case.

### Demo

https://github.com/user-attachments/assets/b444c258-af04-410d-8b5a-96f08e7817fc

### CODE REVIEW

**Comment by lukaszsamson** [comment]
Nice to see my idea taken further and exploring different approaches @jonatanklosko.

My main concern is interaction with corporate firewalls and EDRs that may block socket open/bind. It would be troublesome if running elixir would require security approval on dev machines.

If anybody is interested here's my [Unix domain socket based PoC](https://github.com/lukaszsamson/elixir-socket-lock-poc) I did back in April. The main issue was Windows compatibility. It would require OTP 26 and only stream mode works there - no datagram support yet. See https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/ for details

Still, it would be nice if file locks were available but my [PR adding that](https://github.com/erlang/otp/pull/6251) has been in limbo state since 2022.

**Comment by josevalim** [comment]
@lukaszsamson @jonatanklosko tried the socket approach too but there are bugs in OTP 27.1, so that would require OTP 27.2+. There is also a 100 chars limit on the socket path but we may be able to work around that.

> My main concern is interaction with corporate firewalls and EDRs that may block socket open/bind. 

Do you think this will be the case? If we cannot open up a socket, then it means they cannot start a Phoenix server either?

**Comment by jonatanklosko** [comment]
@lukaszsamson I was concerned with firewalls, because when I tried opening a TCP socket the windows firewall alert would pop up talking about networks access, but I realised it's because it was listening on 0.0.0.0 by default. Listening on loopback avoided the alert, which makes sense, because it skips the network. I don't know about other firewalls though, but then running any local server would be the same issue.

File: lib/mix/lib/mix/lock.ex [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      {:error, _reason} ->
+        :gen_tcp.connect(@loopback, port, @connect_opts)
+    end
+  end
+
+  defp take_over(path, port_path) do
+    lock_path = Path.join(path, "lock_0")
+
+    # We linked to lock_N successfully, so port_path should exist
+    :ok = File.rename(port_path, lock_path)
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:207]:
```suggestion
    File.rename!(port_path, lock_path)
```

File: lib/mix/lib/mix/lock.ex [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+    {:ok, names} = File.ls(path)
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:209]:
```suggestion
    names = File.ls!(path)
```

This applies to all the times you match on `:ok` or `{:ok, _}` when using `File` operations, so I'll stop commenting on those 😄

File: lib/mix/lib/mix/lock.ex [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+    for "port_" <> _ = name <- names do
+      _ = File.rm(Path.join(path, name))
+    end
+
+    for "lock_" <> _ = name <- names, name != "lock_0" do
+      _ = File.rm(Path.join(path, name))
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:216]: Is failing to remove a `lock_` file here an issue? Should we raise in that case?

**Review by whatyouhide — COMMENTED** [review]
Algorithm looks solid on first glance, I left a pass of mostly code-related comments but I'll take a look at this again later 🙃

File: lib/mix/lib/mix/lock.ex [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  # unrelated process and the lock file is stale. Note that it is ok
+  # to use a long timeout, because this scenario is very unlikely.
+  # Theoretically, if an actual lock owner is not able to send the
+  # probe data within the timeout, the lock will fail, however with
+  # a high enough timeout, this should not be a problem in practice.
+
+  @loopback {127, 0, 0, 1}
+  @listen_opts [:binary, ip: @loopback, packet: :raw, nodelay: true, backlog: 128, active: false]
+  @connect_opts [:binary, packet: :raw, nodelay: true, active: false]
+  @probe_data <<"elixirlock">>
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:69]: Doh! :D

File: lib/mix/lib/mix/lock.ex [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+      end
+
+      fun.()
+    after
+      if acquire_lock? do
+        lock = Process.get(pdict_key)
+        unlock(lock)
+        Process.delete(pdict_key)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:101]: Sounds good to me!

File: lib/mix/lib/mix/lock.ex (hunk 3 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:216]: I don't think it would be an issue, but I don't expect a scenario where it would fail, so I think it's a good idea to be more assertive. It made sense for the Unix domain socket implementation, because the files were removed in multiple code paths (though not the lock ones).

File: lib/mix/lib/mix/lock.ex [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  end
+
+  defp lock(path) do
+    :ok = File.mkdir_p(path)
+
+    {:ok, socket} = :gen_tcp.listen(0, @listen_opts)
+    {:ok, port} = :inet.port(socket)
```
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:108]: Good call!

File: lib/mix/lib/mix/lock.ex [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+      File.rm!(Path.join(path, name))
+    end
+
+    for "lock_" <> _ = name <- names, name != "lock_0" do
+      File.rm!(Path.join(path, name))
+    end
+  end
+
+  defp await_close(socket) do
+    {:error, _reason} = :gen_tcp.recv(socket, 0)
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:232]: Warn on unexpected reason?

File: lib/mix/lib/mix/lock.ex [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+  Acquires a lock identified by the given key.
+
+  This function blocks until the lock is acquired by this process,
+  and then executes `fun`, returning its return value.
+
+  This function can also be called if this process already has the
+  lock. In such case the function is executed immediately.
+  """
+  @spec lock(iodata(), (-> term())) :: :ok
+  def lock(key, fun) do
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:82]: Wouldn't `with_lock` or `run_locked` be a better name? `lock` suggests `unlock` will be needed afterwards

Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:82]: I mirrored `Mix.State.lock`, but yeah `with_lock` is much better.

File: lib/mix/lib/mix/lock.ex [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        try_lock(path, socket, port)
+      rescue
+        exception ->
+          # Close the socket to make sure we don't block the lock
+          :gen_tcp.close(socket)
+          reraise exception, __STACKTRACE__
+      end
+    else
+      {:error, reason} ->
+        raise Mix.Error,
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:121]: Call `:gen_tcp.close(socket)` if `:inet.port(socket)` errors

File: lib/mix/lib/mix/lock.ex (hunk 7 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:232]: Anything other than close is very much unexpected, so I think in that case we can close the socket ourselves and retry altogether. I think it would be fine to just assert on `{:error, :closed}`, but doesn't hurt to be cautious.

File: lib/mix/lib/mix/lock.ex [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+      :invalidated ->
+        try_lock(path, socket, port)
+    end
+  end
+
+  defp grab_lock(path, port_path, n) do
+    lock_path = Path.join(path, "lock_#{n}")
+
+    case File.ln(port_path, lock_path) do
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:154]: I'm sure `File.ln` can fail in many more ways. Catch-all `{:error, reason}` and raise?

File: lib/mix/lib/mix/lock.ex [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+  end
+
+  defp probe(port_path) do
+    with {:ok, <<port::unsigned-integer-32>>} when port > 0 <- File.read(port_path),
+         {:ok, socket} <- connect(port) do
+      case :gen_tcp.recv(socket, 0, @probe_timeout_ms) do
+        {:ok, @probe_data} ->
+          {:ok, socket}
+
+        {:error, _reason} ->
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:190]: Surface reason to the caller for easier debugging

File: lib/mix/lib/mix/lock.ex [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+          :gen_tcp.close(socket)
+          :error
+      end
+    else
+      _other -> :error
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:195]: Surface reason to the caller for easier debugging

File: lib/mix/lib/mix/lock.ex [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+    File.write!(port_path, <<port::unsigned-integer-32>>, [:raw])
+
+    case grab_lock(path, port_path, 0) do
+      {:ok, 0} ->
+        # We grabbed lock_0, so all good
+        %{socket: socket, path: path}
+
+      {:ok, _n} ->
+        # We grabbed lock_1+, so we need to replace lock_0 and clean up
+        take_over(path, port_path)
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:138]: Shouldn't we clean up stuff here in case `take_over` fails?

Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:138]: Takeover needs to happen in precise order to avoid race conditions. If anything fails it's ok, we wouldn't get the lock and the next person that will clean up.

File: lib/mix/lib/mix/lock.ex [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+    if has_lock? do
+      fun.()
+    else
+      lock = lock(path)
+      Process.put(pdict_key, true)
+
+      try do
+        fun.()
+      after
+        Process.delete(pdict_key)
```
[the remark this answers is not in the document]
Comment (lukaszsamson) [reply lib/mix/lib/mix/lock.ex:98]: Let's add a note explaining that

File: lib/mix/lib/mix/lock.ex (hunk 13 above)
Comment (lukaszsamson) [reply lib/mix/lib/mix/lock.ex:138]: Let's add a note about cleanup here

File: lib/mix/lib/mix/lock.ex (hunk 7 above)
Comment (lukaszsamson) [reply lib/mix/lib/mix/lock.ex:232]: I wonder if `:gen_tcp` surfaces thread interrupts as `{:error, :eintr}`. Those may need retry

**Comment by lukaszsamson** [comment]
There are a few ways the lock or unlock may fail leaving lock files. Would a `mix unlock` command be needed?

**Comment by jonatanklosko** [comment]
> There are a few ways the lock or unlock may fail leaving lock files. Would a mix unlock command be needed?

Not just lock and unlock, the process may be abruptly terminated at any given point. The locking procedure is designed precisely to handle this and avoid race conditions while doing so. I went through a bunch of other ideas and I believe this one is resilient to failures.

To be more specific about your question, leaving the files behind is not an issue, because another process can check if the file is stale by probing the corresponding port.

File: lib/mix/lib/mix/lock.ex (hunk 9 above)
Comment (whatyouhide) [reply lib/mix/lib/mix/lock.ex:123]: Yes, also.

File: lib/mix/lib/mix/lock.ex (hunk 7 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex]: Not sure, but it is in the function spec, so I added retry in recv and accept.

File: lib/mix/lib/mix/lock.ex [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+  end
+
+  defp listen() do
+    with {:ok, socket} <- :gen_tcp.listen(0, @listen_opts) do
+      case :inet.port(socket) do
+        {:ok, port} ->
+          {:ok, socket, port}
+
+        {:error, reason} ->
+          :socket.close(socket)
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:135]: I think this needs to be

```suggestion
          :gen_tcp.close(socket)
```

File: lib/mix/lib/mix/lock.ex [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+    case listen() do
+      {:ok, socket, port} ->
+        spawn_link(fn -> accept_loop(socket) end)
+
+        try do
+          try_lock(path, socket, port)
+        rescue
+          exception ->
+            # Close the socket to make sure we don't block the lock
+            :gen_tcp.close(socket)
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:118]: The socket is linked to the current process, which I’m assuming would exit due to the exception here. Do we actually need to close the socket manually?

File: lib/mix/lib/mix/lock.ex [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+            reraise exception, __STACKTRACE__
+        end
+
+      {:error, reason} ->
+        raise Mix.Error,
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:123]: I can't see any other instance in the Elixir codebase of manually raising `Mix.Error`. Maybe we want `Mix.raise/1` here?

File: lib/mix/lib/mix/lock.ex [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+              "failed to open a TCP socket while acquiring a lock, reason: #{inspect(reason)}"
+    end
+  end
+
+  defp listen() do
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:128]: We have ~250 instances of 0-arity `def`/`defp` calls with parens, and ~550 without parens. I say we have a winner for the future... 😄 

```suggestion
  defp listen do
```

Super nitpicky, but since we're adding new code 🙃

File: lib/mix/lib/mix/lock.ex [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+
+      {:error, :enoent} ->
+        :invalidated
+
+      {:error, reason} ->
+        raise File.LinkError,
+          reason: reason,
+          action: "create hard link",
+          existing: port_path,
+          new: lock_path
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:193]: Lovely.

File: lib/mix/lib/mix/lock.ex (hunk 10 above)
Comment (whatyouhide) [reply lib/mix/lib/mix/lock.ex:172]: Yep, good callout!

File: lib/mix/lib/mix/lock.ex [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+    end
+  end
+
+  defp accept_loop(listen_socket) do
+    case :gen_tcp.accept(listen_socket) do
+      {:ok, socket} ->
+        _ = :gen_tcp.send(socket, @probe_data)
+        accept_loop(listen_socket)
+
+      {:error, :eintr} ->
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:203]: Note for noobs like me to save them the time to go look this up 😄 

```suggestion
      # eintr is "Interrupted system call".
      {:error, :eintr} ->
```

File: lib/mix/lib/mix/lock.ex [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+        raise File.LinkError,
+          reason: reason,
+          action: "create hard link",
+          existing: port_path,
+          new: lock_path
+    end
+  end
+
+  defp accept_loop(listen_socket) do
+    case :gen_tcp.accept(listen_socket) do
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:198]: If any other error happens here, we'll get a nasty `CaseClauseError`. Some of those are transient (`:system_limit`?), some are not. Maybe we do a simple algorithm of:

1. Accept
2. If `{:error, reason}` and we don't know if reason is transient or not (basically catch all), then sleep for like a second and go back to 1.

With a max number of like 2 or 3 retries for now? I know that just rerunning `mix compile` would likely achieve the same, but it seems like a small price to pay for a potentially-better UX. Thoughts?

File: lib/mix/lib/mix/lock.ex [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+      {:ok, <<0::unsigned-integer-32>>} -> {:error, :ignore}
+      {:ok, <<port::unsigned-integer-32>>} -> {:ok, port}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp connect(port) do
+    # On Windows connecting to an unbound port takes a few seconds to
+    # fail, so instead we shortcut the check by attempting a listen,
+    # which succeeds or fails immediately
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:229]: Oh, nice trick!

File: lib/mix/lib/mix/lock.ex [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+  # unrelated process and the lock file is stale. Note that it is ok
+  # to use a long timeout, because this scenario is very unlikely.
+  # Theoretically, if an actual lock owner is not able to send the
+  # probe data within the timeout, the lock will fail, however with
+  # a high enough timeout, this should not be a problem in practice.
+
+  @loopback {127, 0, 0, 1}
+  @listen_opts [:binary, ip: @loopback, packet: :raw, nodelay: true, backlog: 128, active: false]
+  @connect_opts [:binary, packet: :raw, nodelay: true, active: false]
+  @probe_data "elixirlock"
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:69]: This note might help?

```suggestion
  # The probe data needs to be small enough that it will for sure be
  # sent in a single packet by the socket.
  @probe_data "elixirlock"
```

File: lib/mix/lib/mix/lock.ex [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+
+    names = File.ls!(path)
+
+    for "port_" <> _ = name <- names do
+      File.rm!(Path.join(path, name))
+    end
+
+    for "lock_" <> _ = name <- names, name != "lock_0" do
+      File.rm!(Path.join(path, name))
+    end
```
Comment (whatyouhide) [inline lib/mix/lib/mix/lock.ex:273]: Would this work? Nitpick, because your version works too 🙃 

```elixir
for file <- path |> Path.join("{port_,lock_}*") |> Path.wildcard(),
    Path.basename(file) != "lock_0",
    do: File.rm!(file)
```

**Review by whatyouhide — COMMENTED** [review]
Left another round of comments, fantastic work here. Also took the liberty of turning the algo into a draft of a Mermaid diagram, in case it helps others.

```mermaid
sequenceDiagram
  participant process
  participant port_P as port_P file (for port P)
  participant lock_n
  participant lock_np1 as lock_(n+1)

  process->>process: Open TCP socket on port P
  process->>port_P: Write TCP socket port

  loop Try to acquire lock through port P, starting at n=0
    process->>lock_n: Attempt to link port_P to lock_n
  
    alt Linking succeeds
      port_P->>lock_n: Link
      Note over port_P,lock_n: Process owns the lock

      process->>process: Remove all port_* files
      process->>process: Remove all lock_* files<br/>(except for lock_0)
    else Linking fails
      process->>lock_n: Connect to port in lock_n
      alt Connection succeeds
        lock_n->>process: Socket sends probe data
        Note over lock_n,process: This ensures that the socket is<br/>related to the Mix compilation, not<br/>just a random socket
        process-->>process: Wait for socket to close
        Note over process,process: Someone else has the lock
      else Connection fails (lock is stale)
        process->>lock_np1: Repeat loop with n+1
      end
    end
  end
```

File: lib/mix/lib/mix/lock.ex (hunk 16 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:118]: Theoretically the exception could be rescued, so better to ensure we don't block the lock?

File: lib/mix/lib/mix/lock.ex (hunk 17 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:123]: Sounds good!

File: lib/mix/lib/mix/lock.ex (hunk 15 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:135]: Good catch!

File: lib/mix/lib/mix/lock.ex (hunk 24 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:273]: We want to remove all port_* files first, only then all lock_* files. Not sure if wildcard guarantees the order, but even if it does, I'd rather be explicit about the order of operations.

File: lib/mix/lib/mix/lock.ex (hunk 23 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:69]: Good call :)

**Comment by jonatanklosko** [comment]
@whatyouhide the diagram looks great :D

File: lib/mix/lib/mix/lock.ex (hunk 24 above)
Comment (whatyouhide) [reply lib/mix/lib/mix/lock.ex:277]: It doesn't guarantee that as far as I know. Why do we need to delete all port files first? Might be worth leaving a comment.

File: lib/mix/lib/mix/lock.ex (hunk 21 above)
Comment (whatyouhide) [reply lib/mix/lib/mix/lock.ex:201]: The accepting process is sort of holding its own little lock, so this confuses me a bit:

> it means the client fails to connect and assumes the socket is stale

The client would fail to connect with `:closed` if the socket is closed, not with other errors, no?

File: lib/mix/lib/mix/lock.ex (hunk 24 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:277]: Sidenote: we actually want all files in the directory, so semantically there's no need for wildcard.

> Why do we need to delete all port files first?

This part from the doc comment:

```
# [...] By removing
# all port_P files we make sure that currently paused processes
# that are about to link port_P at lock_N will fail to link, since
# the port_P file will no longer exist (once lock_N is removed).

This is best seen on an example, so imagine this:

```
port_100 (A)
port_200 (B)
lock_0 (stale)
lock_1 (stale)
lock_2 (owned by A)
```

Process A has port 100 and owns lock_2.
Process B has port 200 and is paused right before `File.ln("port_200", "lock_1")`.

Process A first moves port_100 to lock_0. And now wants to remove files.

Let's say process A were to remove lock_1 first. Then process B wakes up and successfully links. At this point Process B will overtake A. Contrarily, if we remove port_200 first, when process B wakes up, it will fail to link.

In other words, once we are sure we are the new owner, we remove port_* files to invalidate any peers attempting to also get the lock.

Does that make sense? :)

Comment (whatyouhide) [reply lib/mix/lib/mix/lock.ex:277]: Yep, makes sense. My head hurts, but it makes sense 😉 Let's maybe leave a comment here too saying that we need to delete these files in the specific order then, wdyt?

**Comment by whatyouhide** [comment]
@jonatanklosko

> the diagram looks great :D

Thanks! Is it *accurate* though? 😄

**Comment by jonatanklosko** [comment]
> Thanks! Is it accurate though?

That's what I meant!

Though after writing the example, I've just realised there's one missing code path where the "[Linking fails]" fails because the source file doesn't exist (the "invalidated" case). In that case we go back to loop with n=0. Similarly after "Someone else has the lock" we also go back to n=0.

File: lib/mix/lib/mix/lock.ex (hunk 21 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:201]: Ah you're right, I think these are not tied. The client is put in the accept queue and at that point they connect successfully, whether `:gen_tcp.accept` grabs it from the queue is separate.

But if the idea is to sleep and retry, this heavily increases the chance that we don't send the probe data to the client on time, which is critical.

File: lib/mix/lib/mix/lock.ex [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+  defp await_probe_data(socket) do
+    case :gen_tcp.recv(socket, 0, @probe_timeout_ms) do
+      {:ok, @probe_data} ->
+        {:ok, socket}
+
+      {:ok, _data} ->
+        :gen_tcp.close(socket)
+        {:error, :unexpected_port_owner}
+
+      {:error, :eintr} ->
```
Comment (lukaszsamson) [inline lib/mix/lib/mix/lock.ex:251]: I did some research and it seems this is not needed. It's not mentioned in the manual nor Was I able to find any code explicitly checking for that result.

File: lib/mix/lib/mix/lock.ex [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+  @doc """
+  Acquires a lock identified by the given key.
+
+  This function blocks until the lock is acquired by this process,
+  and then executes `fun`, returning its return value.
+
+  This function can also be called if this process already has the
+  lock. In such case the function is executed immediately.
+  """
+  @spec with_lock(iodata(), (-> term())) :: :ok
```
Comment (sabiwara) [inline lib/mix/lib/mix/lock.ex:83]: I think it should be:

```suggestion
  @spec with_lock(iodata(), (-> term())) :: term()
```

File: lib/mix/lib/mix/lock.ex (hunk 25 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:251]: If I understand correctly the C code for Unix recv is [here](https://github.com/erlang/otp/blob/39383439d05b2135acb498d3e4557b70f16714f0/erts/emulator/nifs/unix/unix_socket_syncio.c#L2836-L2851). There are no checks for `EINTR`, and the default error code path leads to returning `erl_errno_id(saveErrno)`, which does recognise `EINTR`, so it seems to me that technically it could be returned.

I tried to test it by setting SIGALRM handler and sending the signal while `:gen_tcp.recv` is blocked, and that didn't interrupt the call. But then I realised it's probably that the signal is received to the main thread while recv runs in another.

File: lib/mix/lib/mix/lock.ex (hunk 23 above)
Comment (jonatanklosko) [reply lib/mix/lib/mix/lock.ex:71]: Actually, if we pass the desired size to `recv`, it will accumulate across packets, so that should be better :D

File: lib/mix/lib/mix/tasks/compile.all.ex [hunk 27]
```diff
@@ -12,6 +12,13 @@ defmodule Mix.Tasks.Compile.All do
   @impl true
   def run(args) do
     Mix.Project.get!()
+
+    Mix.Lock.with_lock(Mix.Tasks.Compile.compile_lock_key(), fn ->
+      do_run(args)
+    end)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/compile.all.ex:18]: Let's encapsulate this in `Mix.Task.Compiler.with_lock(fn -> ... end)` instead of passing `compile_lock_key` around. It can be `@doc false` for now but we will eventually want to make it public. :)

Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.all.ex:18]: Or even better, let's pass `config` as argument to the function so we fetch it less often?

Comment (josevalim) [reply lib/mix/lib/mix/tasks/compile.all.ex:18]: Or maybe it should be in `Mix.Project.with_build_lock`.

**Review by josevalim — APPROVED** [review]
Fantastic work! :heart:

I added only one nitpick, about encapsulating part of the lock either in `Mix.Tasks.Compiler` (or perhaps in `Mix.Project`, such as `Mix.Project.with_build_lock`, since we also use it for deps.compile).

The only other comment is if we should print a message that we are waiting for another process. I know that Rust's `cargo`s does this. In this case, we can maybe store the PID in the file and use it as part of the message. If we don't want to print by default, we should at least print on `MIX_DEBUG`.

In the future we should also have a lock related to `mix deps.get` and it may need to be a read-write lock. :) We can chat later!

**Comment by jonatanklosko** [comment]
@josevalim I want with `Mix.Project.with_build_lock/2`. I also added the message, so this gets printed:

```
Waiting for lock on the build directory (held by process 94019)
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
